src/systems/shadow-system.js

```diff
@@ -64,10 +64,11 @@ function resizeShadowCameraFrustum(light, boundingBox) {
   verts[7].set(max.x, max.y, max.z);
 
   light.updateMatrices();
-  inverseLightMatrixWorld.copy(light.matrixWorld).invert();
+  light.shadow.updateMatrices(light);
```

Comment:
Hmm. Would you have to do this in upstream THREE? If not I bet its a result of our matrix dirty flags patch, so we should probably include fixing lights in that patch.

Reply:
> Hmm. Would you have to do this in upstream THREE?

I don't think so, maybe. Shadow camera matrices are updated in `WebGLShadowMap` when rendering shadow depth to shadow map. It would be good for most of the use cases because generally light shadow camera matrices are not used in user code.

I guess the reason why the shadow camera matrices update is separated from light matrices update is, in addition to optimization, shadow camera matrices can be affected from `light.target` which can be anywhere in the scene graph (or can be out side of the scene graph). It doesn't greatly fit to the current `.updateMatrixWorld()` so I speculate Three.js wants to calculate shadow camera matrices after updateing the entire scene graph matrices.

> I bet its a result of our matrix dirty flags patch

I don't think so. Light matrices and shadow camera matrices are different. The latter but the latter taking account of looking at `light.target` but the former doesn't.

Reply:
Ah. Then I am a bit confused why this is needed in Hubs but not upstream then. In GLTFLoader the lights target is added as a child and set up such that the light will be facing down its -z axis https://github.com/mrdoob/three.js/blob/dev/examples/jsm/loaders/GLTFLoader.js#L522

Reply:
Again, `light.target` can be anywhere in the scene graph in Three.js.

https://threejs.org/docs/#api/en/lights/DirectionalLight.target

Then Three.js wants to calculate shadow camera after the entire scene matrices are updated so it won't update in object's `.updateMatrixWorld()` but updates it in renderer after entire scene matrices are up to date. I don't think object `.updateMatrixWorld()` should update the shadow camera in the upstream.

Reply:
Ah yes. What I kept missing is that we are reading the shadow camera's matrix here, so it needs to be up to date. So this is not about the shadow camera's matrix not getting computed correctly (THREE will update it correctly during render) but about *when* it is updated. In most cases you don't care about it being updated until it is about to be used during render. But in our case we care about it being updated here since we are using it to transform points to figure out the bounding box.